subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionPlan.java

```diff
@@ -484,36 +482,98 @@ public TaskInfo getTaskToExecute() {
+                final AtomicBoolean restartIteration = new AtomicBoolean();
+                final Set<String> lockedProjects = Sets.newHashSet();
                 while (iterator.hasNext()) {
-                    TaskInfo taskInfo = iterator.next();
-                    allTasksComplete = allTasksComplete && taskInfo.isComplete();
-                    if (taskInfo.isReady() && taskInfo.allDependenciesComplete() && canRunWithWithCurrentlyExecutedTasks(taskInfo)) {
-                        nextMatching = taskInfo;
-                        iterator.remove();
-                        break;
+                    final TaskInfo taskInfo = iterator.next();
+                    Project project = taskInfo.getTask().getProject();
+                    final String projectPath = ((ProjectInternal) project).getIdentityPath().toString();
+                    if (taskInfo.isReady() && taskInfo.allDependenciesComplete()) {
+                        try {
+                            final ProjectLock projectLock = projectLockService.getProjectLock(projectPath);
+                            boolean couldNotLock = !projectLock.tryWithProjectLock(new Runnable() {
+                                @Override
+                                public void run() {
+                                    boolean canExecute = false;
+                                    try {
+                                        if (lockedProjects.contains(projectLock.getProjectPath())) {
+                                            // If the state of project locks has changed, we want to restart
+                                            // task selection so that we can be sure that the earliest task
+                                            // from this project in the execution queue is selected first.
+                                            restartIteration.set(true);
+                                        } else {
+                                            if (canRunWithWithCurrentlyExecutedTasks(taskInfo)) {
+                                                selected.set(true);
+                                                iterator.remove();
+                                                if (taskInfo.allDependenciesSuccessful()) {
+                                                    taskInfo.startExecution();
+                                                    recordTaskStarted(taskInfo);
+                                                    canExecute = true;
+                                                } else {
+                                                    taskInfo.skipExecution();
+                                                }
+                                            }
+                                        }
+                                    } finally {
+                                        // We need to unlock the plan before unlocking the project.  This prevents a deadlock
```

Comment:
I like this a lot.  +100 to anything that simplifies the management of these locks and prevents us from having to have elaborate rules about the relationships between locks.

One thing this does imply is that we would have asynchronous methods on the project lock, which we were initially trying to avoid, but I don't think this is a terrible thing.  There is an important trade off which is that we no longer need a blocking call to lock the project.  The non-blocking call seems important to avoiding deadlocks and more than makes up for any overhead the asynchronous methods might introduce.

I mocked up some (terrible) pseudo-code to wrap my head around this - let me know if this aligns with what you were thinking:

```
task worker -> TaskExecutionPlan.executeWithTask
    executeWithTask -> CoordinationService.withStateLock
        withStateLock -> acquireStateLock
        withStateLock -> selectTaskAction
            selectTaskAction -> iterate until task is found
            selectTaskAction -> ProjectLock.tryProjectLock
            selectTaskAction -> if locked, remove task from execution queue (selected)
        withStateLock -> releaseStateLock
        withStateLock -> if task selected, return, else wait until state changes and repeat

    executeWithTask -> execute taskExecution
        taskExecution -> CoordinationService.withStateLock
            withStateLock -> acquireStateLock
            withStateLock -> acquireWorkerLeaseAction
            withStateLock -> releaseStateLock
        taskExecution -> execute task action
        taskExecution -> AsyncWorkTracker.waitForCompletion
            waitForCompletion -> CoordinationService.withStateLock
                withStateLock -> acquireStateLock
                withStateLock -> releaseProjectLockAction
                    releaseProjectLockAction -> ProjectLock.unlock 
                        unlock -> CoordinationService.notifyStateChange
                withStateLock -> releaseStateLock
            waitForCompletion -> wait for all async work items
            waitForCompletion -> CoordinationService.withStateLock
                withStateLock -> acquireStateLock
                withStateLock -> acquireProjectLockAction
                    acquireProjectLockAction -> ProjectLock.tryProjectLock
                    acquireProjectLockAction -> if not locked, release worker leases
                    acquireProjectLockAction -> if locked, reacquire worker leases (if released)
                withStateLock -> releaseStateLock
                withStateLock -> if project not locked, wait until state changes and repeat
        taskExecution -> if more task actions, repeat
        taskExecution -> CoordinationService.withStateLock
            withStateLock -> acquireStateLock
            withStateLock -> releaseWorkerLeaseAction
            withStateLock -> releaseStateLock

    executeWithTask -> CoordinationService.withStateLock
        withStateLock -> acquireStateLock
        withStateLock -> releaseProjectLockAction
            releaseProjectLockAction -> ProjectLock.unlock
                unlock -> CoordinationService.notifyStateChange
        withStateLock -> releaseStateLock
    executeTask -> if all tasks complete, return false, else return true
task worker -> if true, repeat, else finished
```